<a href="https://colab.research.google.com/github/SergeiSa/Computational-Intelligence-Slides-Spring-2021/blob/main/Google%2520Colab%2520notebooks/CI_LyapunovEq.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lyapunov equations

Lyapunov equations for continious systems has form:

$AP + PA^{\top}  =-Q$

and for discrete systems it is 

$A P A^{\top} - P = -Q$

As long as there exists such positive definite $P$ that Lyapunov equations holds for a positive definite $Q$, the system is stable.

Let's see it in code:

In [ ]:
import numpy as np
from scipy.linalg import solve_continuous_lyapunov
from scipy.linalg import solve_discrete_lyapunov
from scipy.linalg import eig


Q = np.array([[-1, 0], [0, -1]])

A = np.array([[-10, 5], [-5, -10]])
e, v = eig(A)
print("eig(A)", e)

P = solve_continuous_lyapunov(A, Q)
print("P", P)
e, v = eig((A.transpose().dot(P) + P.dot(A)))
print("eig(A'P + P*A)", e)
print(" ")
print(" ")


A = np.array([[0.9, 0.5], [-0.2, -0.8]])
e, v = eig(A)
print("eig(A)", e)

P = solve_discrete_lyapunov(A, Q)
print("P", P)
print("(A'PA - P + Q ):")
print(((A.dot(P)).dot(A.transpose()) - P + Q))


eig(A) [-10.+5.j -10.-5.j]
P [[ 5.00000000e-02  7.34706413e-20]
 [-1.24900090e-18  5.00000000e-02]]
eig(A'P + P*A) [-1.+0.j -1.+0.j]
 
 
eig(A) [ 0.83898669+0.j -0.73898669+0.j]
P [[-4.03347296  0.9268445 ]
 [ 0.9268445  -2.40207966]]
(A'PA - P + Q ):
[[0.00000000e+00 3.33066907e-16]
 [1.11022302e-16 4.44089210e-16]]


Test the following systems' stability:



$$x_{i+1} = 
\begin{pmatrix} 1.5 & 0.2 \\ -0.15 & 0.23
\end{pmatrix}
x_i
$$


$$x_{i+1} = 
\begin{pmatrix} -1 & -1 \\ -2 & 0.1
\end{pmatrix}
x_i
$$


$$x_{i+1} = 
\begin{pmatrix} -3 & -1 \\ -1.5 & -10.3
\end{pmatrix}
x_i
$$


$$x_{i+1} = 
\begin{pmatrix} -0.2 & -1 \\ 1.7 & 1.1
\end{pmatrix}
x_i
$$

Test stability of continious systems with the same state matrices:


$$\dot x = 
\begin{pmatrix} 1.5 & 0.2 \\ -0.15 & 0.23
\end{pmatrix}
x
$$


$$\dot x = 
\begin{pmatrix} -1 & -1 \\ -2 & 0.1
\end{pmatrix}
x
$$


$$\dot x = 
\begin{pmatrix} -3 & -1 \\ -1.5 & -10.3
\end{pmatrix}
x
$$


$$\dot x = 
\begin{pmatrix} -0.2 & -1 \\ 1.7 & 1.1
\end{pmatrix}
x
$$

Consider the following matrices:

$$\dot x = 
\begin{pmatrix} -10.05 & -0.021 & -0.02 \\  
0 & 0 & 0 \\  
-0.022 & 0.0032 & -10.055
\end{pmatrix}
x
$$

$$\dot x = 
\begin{pmatrix} -2.009 & 0 & 0.0012 \\  
0.05 & 0 & -0.041 \\  
-0.042 & 0 & -4.055
\end{pmatrix}
x
$$

Can you reason about assymptotic stability of the system without computations?

**Answer:**

First system is Lyapunov stable, but not assymptotically stable. You can see that its eigenvalues are not going to have negative eigenvalues, but also you see that the matrix has a non-trivial null space.

You can also directly notice that the second component of $x$ is not going to change, rulling out assymptotic stability.

Second sistem has a non-trivial null space too, meaning some of its eigenvalues are going to be 0. 

Consider the following system again:

$$\dot x = 
\begin{pmatrix} -10.05 & -0.021 & -0.02 \\  
0 & 0 & 0 \\  
-0.022 & 0.0032 & -10.055
\end{pmatrix}
x
$$

You know that $x_2$ is a constant. Let's denote the value of $x_2$ as $c$. Can you rewrite the system in terms of remaining variables $x_1$ and $x_3$ and prove its stability using the Lyapuniv eq?

## Solution

### Way 1

Denote:

$$y = 
\begin{pmatrix} 
x_1 \\  
x_3
\end{pmatrix}
$$

And rewrite:

$$\dot y = 
\begin{pmatrix} -10.05 & -0.02 \\  
-0.022 & -10.055
\end{pmatrix}
y
+
\begin{pmatrix} -0.021 \\  
0.0032
\end{pmatrix}
c
$$

The rest you know how to do.

### Way 2

Orthonormal basis in the column space of the state matrix in this case is:

$$
C =
\begin{pmatrix} 
1 & 0 \\  
0 & 0 \\  
0 & 1
\end{pmatrix}
$$

Motion of the system takes place in that column space. Let's denote $y = C^{\top}x$, and as long as $x$ is in this column space, it is true that $x = Cy$. But if $x$ is not in the column space, it is $x = Cy + x^*$

Notice that $x^*$ is in the left null space of the state matrix, as long as $y = C^{\top}x$, because: 

$$Cy = CC^{\top}x$$
$$x-x^* = CC^{\top}x$$
$$x^* = (I-CC^{\top})x$$

where the last expression is a projection of $x$ onto the left null space of the state matrix. Orthonormal basis in the left null space of the state matrix is:

$$
L =
\begin{pmatrix} 
0 \\  
1 \\  
0
\end{pmatrix}
$$

And we know that $x_2 = c$, so $x^* = Lc$.

Variable $y$ is the new coordinates in the column space basis.

Let's project teh dynamics into the column space. First we multiply it by $C^{\top}$:

$$C^{\top} \dot x = 
C^{\top}
\begin{pmatrix} -10.05 & -0.021 & -0.02 \\  
0 & 0 & 0 \\  
-0.022 & 0.0032 & -10.055
\end{pmatrix}
x
$$

$$\dot y = 
\begin{pmatrix} -10.05 & -0.021 & -0.02 \\ 
-0.022 & 0.0032 & -10.055
\end{pmatrix}
x
$$

Then, since $x = Cy + Lc$ on teh system trajectory, we get:

$$\dot y = 
\begin{pmatrix} -10.05 & -0.021 & -0.02 \\ 
-0.022 & 0.0032 & -10.055
\end{pmatrix}
(Cy + Lc)
$$

$$\dot y = 
\begin{pmatrix} -10.05 & -0.02 \\ 
-0.022 & -10.055
\end{pmatrix}
y
+
\begin{pmatrix} 
-0.021 \\ 
0.0032
\end{pmatrix}
c
$$

From here you apply Lyapunov eq. directly.